// @ilc-technology/form-renderer/fields/Checkbox.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var dom_1 = require("../utils/dom");
var Field_1 = __importDefault(require("./Field"));
var misc_1 = require("../utils/misc");
var Checkbox = (function (_super) {
    __extends(Checkbox, _super);
    function Checkbox(configuration, form) {
        var _this = _super.call(this) || this;
        _this.configuration = configuration;
        _this.form = form;
        _this._previousValue = false;
        _this.render = function () {
            var _a = _this, configuration = _a.configuration, form = _a.form;
            var name = configuration.name, placeholder = configuration.placeholder, label = configuration.label;
            var applyGud = form.settings.applyGud;
            var defaultValue = configuration.defaultValue || false;
            var initialValue = {
                checked: defaultValue ? defaultValue : undefined,
            };
            _this._inputElement = dom_1.createElement('input', __assign({ id: _this.form.settings.id + "-" + configuration.name, type: 'checkbox', name: name,
                placeholder: placeholder, class: 'check-custom ef-boolean__input' }, initialValue));
            _this._previousValue = defaultValue;
            _this.form.data[configuration.name] = defaultValue;
            var span = dom_1.createElement('span', {
                for: _this.form.settings.id + "-" + configuration.name,
                class: 'check-squared ef-boolean__element -checkbox',
            });
            var checkboxCustom = dom_1.createElement('label', {
                for: _this.form.settings.id + "-" + configuration.name,
                class: 'label-text',
            }, _this._inputElement, span);
            var abbr = dom_1.createElement('abbr', {
                class: 'ef-form-label__required',
                title: 'required',
            }, ' *');
            var actualLabel = label ? label : '';
            var labelText = dom_1.createElement('span', {
                for: form.settings.id + "-" + configuration.name,
            }, actualLabel || '');
            var labelNode = dom_1.createElement('label', {
                for: _this.form.settings.id + "-" + configuration.name,
                class: 'ef-boolean__label',
            }, labelText, applyGud && _this.isRequired() ? abbr : '');
            var errorMessageWrapper = dom_1.createElement('div', {
                class: 'ef-form-validation -is-invalid ef-form-error-message',
            });
            _this.htmlElement = dom_1.createElement('div', {
                class: "ef-form-control ef-form-" + configuration.name + "-wrapper\n        ef-form-" + configuration.type + "-wrapper ef-input-w ef-boolean u-s-mb-xs",
            }, checkboxCustom, labelNode, errorMessageWrapper);
            _this._inputElement.addEventListener('change', function (_a) {
                var target = _a.target;
                return __awaiter(_this, void 0, void 0, function () {
                    var checked, validation;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                checked = target.checked;
                                this.form.data[configuration.name] = checked;
                                _super.prototype._executeListeners.call(this, 'change');
                                return [4, _super.prototype.validate.call(this)];
                            case 1:
                                validation = _b.sent();
                                dom_1.createErrorMessageNodes(errorMessageWrapper, validation.errorList);
                                _super.prototype._executeListeners.call(this, 'validationCompleted');
                                this._previousValue = checked;
                                return [2];
                        }
                    });
                });
            });
            return _this.htmlElement;
        };
        return _this;
    }
    Checkbox.builder = function (conf, form) {
        return new Checkbox(conf, form);
    };
    Object.defineProperty(Checkbox.prototype, "value", {
        get: function () {
            return this._inputElement ? this._inputElement.checked : false;
        },
        set: function (value) {
            if (this._inputElement) {
                this._inputElement.checked = value;
                this.form.data[this.configuration.name] = value;
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Checkbox.prototype, "previousValue", {
        get: function () {
            return this._previousValue;
        },
        enumerable: false,
        configurable: true
    });
    Checkbox.prototype.triggerChangeEvent = function () {
        if (this._inputElement) {
            misc_1.triggerEvent(this._inputElement, 'change');
        }
    };
    return Checkbox;
}(Field_1.default));
exports.default = Checkbox;
Field_1.default.register('checkbox', Checkbox.builder);
